### PR TITLE
MINOR: Optimization should only run with more than one repartition topic

### PR DESCRIPTION
**Pull request by bbejeck**
If there is only one repartition topic as the result of a key-change operation, then the optimization should not take place as the resulting sub-topology will have the same structure.  While not strictly an issue I'm very much in favor of not performing unnecessary operations. 

I've added a test which fails without the update to the optimization.  The existing test suite ensures all other code not impacted by the change. 

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage, and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by bbejeck** [comment]
calling @guozhangwang, @mjsax, and @vvcephei for reviews

**Comment by mjsax** [comment]
Java11: https://builds.apache.org/job/kafka-pr-jdk11-scala2.12/544/#showFailuresLink
```
Test Result (12 failures / +2)

kafka.admin.ResetConsumerGroupOffsetTest.testResetOffsetsNotExistingGroup
kafka.api.PlaintextConsumerTest.testLowMaxFetchSizeForRequestAndPartition
kafka.api.SaslSslAdminClientIntegrationTest.testAclDescribe
kafka.api.SaslSslAdminClientIntegrationTest.testLegacyAclOpsNeverAffectOrReturnPrefixed
kafka.api.TransactionsTest.testFencingOnTransactionExpiration
org.apache.kafka.streams.integration.RepartitionWithMergeOptimizingIntegrationTest.shouldSendCorrectRecords_OPTIMIZED
org.apache.kafka.streams.integration.ResetIntegrationTest.shouldNotAllowToResetWhenInputTopicAbsent
org.apache.kafka.streams.integration.ResetIntegrationWithSslTest.testReprocessingFromScratchAfterResetWithIntermediateUserTopic
org.apache.kafka.streams.integration.SuppressionIntegrationTest.shouldSuppressIntermediateEventsWithRecordLimit
kafka.api.PlaintextConsumerTest.testAutoCommitIntercept
kafka.api.CustomQuotaCallbackTest.testCustomQuotaCallback
kafka.tools.MirrorMakerIntegrationTest.testCommaSeparatedRegex
```

Java8:
```
kafka.api.SaslPlainPlaintextConsumerTest.testCoordinatorFailover
kafka.api.SaslSslConsumerTest.testCoordinatorFailover
```

**Review by mjsax — APPROVED** [review]
LGTM. Call for second review @vvcephei -- can merge afterwards.

**Comment by mjsax** [comment]
Retest this please.

**Comment by vvcephei** [comment]
Reviewing...

**Comment by vvcephei** [comment]
Retest this, please.

**Comment by vvcephei** [comment]
Hey @bbejeck ,

Thanks for this. It LGTM!

Just a question about the compatibility story...

This means that if I *don't* name my repartitioning nodes (i.e., using Grouped etc), then I have to reset my app to upgrade from 2.1 to 2.2, right?

If I *do* name the repartitioning nodes, I should be fine to upgrade, though, right?

Thanks,
-John

**Comment by vvcephei** [comment]
Hey @bbejeck , I thought about this more... I'm wondering if there might be a bad interaction with the optimizations that will be released in 2.1 .

Imagine I have a topology like this:
1. selectKey
2. filter
3. groupByKey(named)
4. count

In 2.1, the optimizer would move the repartition topic from just after 3 to just after 1. If we release this PR in 2.2, then upon upgrading the optimizer would decide *not* to move the repartition topic, so it would stay just after 3. However, if I was previously running the app on 2.1 and didn't do a reset, that means that the repartition topic effectively moves from after 1 to after 3 when I upgrade.

This would be bad, since the repartition topic actually contains some records, so when it moves, it can effectively bypass the filter.

I'm not sure we can currently detect this situation, so it may actually not be safe to use the code in this PR, unless we can get it included in the 2.1 release as well.

Does this make sense?

**Comment by bbejeck** [comment]
@vvcephei that is a good point, although just for clarity Kafka Streams places the repartition topic before the `groupBy`, that is beside the point.

Although I think this could be a rare edge condition, I'm inclined to agree with you that if we can't get this change into `2.1` we don't merge this PR.  

My main concern is users having to reset the application when moving to `2.1` with optimizations and having to reset the application again in `2.2`, but if we add other optimizations and they opt in another reset is possible.

\cc @guozhangwang @mjsax  WDYT?

**Comment by vvcephei** [comment]
@bbejeck : Agreed. On that last point, this would be alleviated if we also introduce a new "2.2 optimizations" config value and put this code behind it, so if users don't change their configs but upgrade to 2.2, the optimizer will still behave the same.

**Comment by mjsax** [comment]
I don't see why a reset would be required? In 2.1 with optimization disable, the repartition topic will be after filter(). In 2.1, with optimization enabled, the repartition topic will be after filter(). and in 2.2 with this change, the repartition topic will still be after filter().

What do I miss?

**Comment by vvcephei** [comment]
@mjsax , I think the troublesome scenario is:
* run with optimizations *enabled* on 2.1
* upgrade to 2.2 without changing the optimization setting.

This would cause the repartition topic to move from before the filter to after it, potentially teleporting whatever records happen to be enqueued past the filter.

**Comment by bbejeck** [comment]
@vvcephei there's one additional thing to consider here that may mitigate this circumstance.  We aggressively purge data from repartition topics, so the chance of having records left over in the repartition topic after a redeployment/stop and start is quite small.

**Comment by mjsax** [comment]
@vvcephei Can you explain why the optimization topic would move before the filter? From my understanding, it would still be after the filter...

In the pre-optimized topology the `groupByKey` is a "virtual node" and there is no `Processor`, ie, `groupByKey` represent the repartition topic if the is any or is "removed" in the `ProcessorTopology` entirely. Basically, when `groupByKey()` is executed, a "repartition step" (if required), is added after the `filter()` node.

In 1.1, we would generate multiple repartition topics, and each downstream operator would consume data from a different repartition topic.
```
filterNode --+--> repartition --> node1
             |
             +--> repartition --> node2
```
In 2.0, with optimization enable, we rewrite the graph to
```
filterNode --> repartition --+--> node1
                             |
                             +--> node2
```

From my understanding, this change only "skips" the merge step if there in only one "repartition" node, ie, it is only affective, if the input topology is:
```
filterNode --> repartition --> node1
```
avoiding an unnecessary "rewrite" to
```
filterNode --> repartition --> node1
```

Or is my understanding incorrect?

@bbejeck I don't think we should talk chances. This would be race condition that we should not introduce.

**Comment by bbejeck** [comment]
>@bbejeck I don't think we should talk chances. This would be race condition that we should not introduce.

Yeah, you are right about that.
 
@mjsax regarding the repartition node movement, when we optimize the graph we move the repartition node to be the first child of the key-changing operation so the following topology
```
source--+--> select-key --+--> filterNode --+--> repartition --> node1
                                            |
                                            +--> repartition --> node2
```
get rewritten to 
```
source--+--> select-key--+--> repartition --+--> filterNode --+--> node1
                                                              |
                                                              +--> node2
```
So @vvcephei's concerns https://github.com/apache/kafka/pull/5894#issuecomment-439954251 and 
https://github.com/apache/kafka/pull/5894#issuecomment-438403969 are correct.

If it's not too late to get this PR cherry-picked to `2.1` then maybe we can merge it, otherwise I can just close this PR

**Comment by vvcephei** [comment]
@bbejeck Dong just sent the email that the release vote passed, so I think it's too late (if I understand the project policies). :/

**Comment by bbejeck** [comment]
>@bbejeck Dong just sent the email that the release vote passed, so I think it's too late (if I understand the project policies). :/

Yes, I believe the "ship has sailed"

**Comment by mjsax** [comment]
Thanks for clarification @bbejeck. I guess, for the concrete example, moving the topic might actually have a negative impact. We need to pipe more data through the repartition topic if we place it before the filter. To be fair, it sounds like a complex problem to resolve.

Overall, due to compatibility concerns, I agree that we should not merge this, if we don't have an answer for it. Because of my example from above, it seems worth to explore though (maybe in a broader context, ie, an optimization on "where to place repartition topics" if we have some freedom to move them around).

Thoughts?

**Comment by bbejeck** [comment]
@mjsax thanks for the reply.

I'm not sure I agree that we have "freedom "to move reparation topics around.  
I have a strong opinion that when performing an optimization, the repartition topic should be an immediate child of the key-changing operation.  Of course, we can always have more discussion on the matter.

**Pull request closed without merging**
